### DISCUSSION TITLE
How to export user defined data types? && Ideas to create Typedef in archive with scripting?

### DISCUSSION THREAD

**Discussion by SparkleStar88** [NationalSecurityAgency/ghidra#4360] [Q&A]
Typically, I want to store all user-defined data types for the future, so I create a new file archive as a storage container.
```		
FileArchiveBasedDataTypeManager FDM = FileDataTypeManager.createFileArchive(archiveFile);
```
Then I use this API 
```
FDM.addDataTypes(collection<DataType>, DataTypeConflictHandler.DEFAULT_HANDLER, monitor);
```
to add all target data types into the data type manager.

After that, I tried to print out all the data types inside the manager, it worked as desired. However, it seems that the archive is not synced with its backed-up data type manager FDM. When I open that archive again, there is nothing inside.

So are the data types for the manager stored separated from the file archive? Anyone has ideas?

Also, is there any way to use scripting to simulate the following operation?
![Screenshot 2022-06-19 120604](https://user-images.githubusercontent.com/56299052/174465391-0522eeca-a6ff-4e1b-a0eb-dc3d4d45c440.png)

**Comment by dragonmacher** [comment]
Regarding the types not being available, It sounds like you did not call `save()` on your manager after you were done.

Regarding create a data type, such as a typedef, as shown in the picture above, you can just create the type and add it to you manager, such as:
```
	DataType typedef = new TypedefDataType(categoryPath, name, dataType);
	DataType newdt = dataTypeManager.addDataType(typedef, plugin.getConflictHandler());

```

To show that type in and editor in the UI, you should then be able to do this:
```
	DataTypeManagerService service = state.tool.getService(DataTypeManagerService.class);
	service.edit(baseDataType);
```
